provider/openstack/firewaller.go

```
@@ -545,92 +545,139 @@ func (c *neutronFirewaller) setUpGlobalGroup(groupName string, apiPort int) (neu
+		return zeroGroup, errors.Errorf("more than one security group with name: %s", name)
 	}
-	// Doesn't exist, so try and create it.
-	newGroup, err := neutronClient.CreateSecurityGroupV2(name, "juju group")
-	if err != nil {
-		return zeroGroup, err
+
+	have := newRuleInfoSetFromRules(group.Rules)
+	want := newRuleInfoSetFromRuleInfo(rules)
+
+	// Find rules we want to delete, that we have but don't want, and
+	// delete them.
+	remove := make(ruleInfoSet)
+	for k := range have {
+		// Neutron creates 2 egress rules with any new Security Group.
+		// Keep them.
+		if !want[k].found && k.Direction != "egress" {
+			remove[k] = have[k]
+		}
 	}
+	for _, v := range remove {
+		_ = neutronClient.DeleteSecurityGroupRuleV2(v.ruleId)
+	}
+
+	// Find rules we want to add, that we want but don't have, and add
+	// them.
+	add := make(ruleInfoSet)
+	for k := range want {
+		if !have[k].found {
+			add[k] = have[k]
+		}
+	}
+
 	// The new group is created so now add the rules.
-	for _, rule := range rules {
-		rule.ParentGroupId = newGroup.Id
+	for rule, _ := range add {
+		rule.ParentGroupId = group.Id
 		// Neutron translates empty RemoteIPPrefix into
 		// 0.0.0.0/0 or ::/0 instead of ParentGroupId
 		// when EthernetType is set
 		if rule.RemoteIPPrefix == "" {
-			rule.RemoteGroupId = newGroup.Id
+			rule.RemoteGroupId = group.Id
 		}
-		groupRule, err := neutronClient.CreateSecurityGroupRuleV2(rule)
+		_, err := neutronClient.CreateSecurityGroupRuleV2(rule)
 		if err != nil {
 			return zeroGroup, err
 		}
-		newGroup.Rules = append(newGroup.Rules, *groupRule)
 	}
-	return *newGroup, nil
-}
 
-func countIngressRules(rules []neutron.SecurityGroupRuleV2) int {
-	count := 0
-	for _, rule := range rules {
-		if rule.Direction == "ingress" {
-			count += 1
-		}
+	// Since we may have done a few add or delete rules, get a new
+	// copy of the security group to return containing the end
+	// list of rules.
+	groupsFound, err = neutronClient.SecurityGroupByNameV2(name)
```

we're ignoring the error here too